File: ios/algorithms/HMACCrypto.m
#import "HMACCrypto.h"
#import "CryptoUtils.h"
#import <CommonCrypto/CommonHMAC.h>

@implementation HMACCrypto

+ (nullable NSString *)hmacSha256Hex:(NSString *)dataHex keyHex:(NSString *)keyHex {
    NSData *contentData = [CryptoUtils fromHex:dataHex];
    NSData *keyData = [CryptoUtils fromHex:keyHex];
    
    NSData *hmacData = [self computeHmac:contentData key:keyData algorithm:HMACAlgorithmSHA256];
    if (hmacData) {
        return [CryptoUtils bytesToHex:hmacData];
    }
    return nil;
}

+ (nullable NSData *)computeHmac:(NSData *)data key:(NSData *)key algorithm:(HMACAlgorithm)algorithm {
    CCHmacAlgorithm ccAlgorithm;
    NSUInteger digestLength;
    
    switch (algorithm) {
        case HMACAlgorithmSHA1:
            ccAlgorithm = kCCHmacAlgSHA1;
            digestLength = CC_SHA1_DIGEST_LENGTH;
            break;
        case HMACAlgorithmSHA224:
            ccAlgorithm = kCCHmacAlgSHA224;
            digestLength = CC_SHA224_DIGEST_LENGTH;
            break;
        case HMACAlgorithmSHA256:
            ccAlgorithm = kCCHmacAlgSHA256;
            digestLength = CC_SHA256_DIGEST_LENGTH;
            break;
        case HMACAlgorithmSHA384:
            ccAlgorithm = kCCHmacAlgSHA384;
            digestLength = CC_SHA384_DIGEST_LENGTH;
            break;
        case HMACAlgorithmSHA512:
            ccAlgorithm = kCCHmacAlgSHA512;
            digestLength = CC_SHA512_DIGEST_LENGTH;
            break;
        default:
            return nil;
    }
    
    NSMutableData *hash = [NSMutableData dataWithLength:digestLength];
    CCHmac(ccAlgorithm, key.bytes, key.length, data.bytes, data.length, hash.mutableBytes);
    
    return hash;
}

+ (nullable NSString *)hmacBase64:(NSString *)dataBase64 keyBase64:(NSString *)keyBase64 algorithm:(HMACAlgorithm)algorithm {
    NSData *dataBytes = [CryptoUtils decodeBase64:dataBase64];
    NSData *keyBytes = [CryptoUtils decodeBase64:keyBase64];
    
    if (!dataBytes || !keyBytes) {
        return nil;
    }
    
    NSData *hmacBytes = [self computeHmac:dataBytes key:keyBytes algorithm:algorithm];
    if (hmacBytes) {
        return [CryptoUtils encodeBase64:hmacBytes];
    }
    return nil;
}

+ (nullable NSString *)hmacStringBase64:(NSString *)data keyBase64:(NSString *)keyBase64 algorithm:(HMACAlgorithm)algorithm {
    NSData *dataBytes = [CryptoUtils stringToUtf8Data:data];
    NSData *keyBytes = [CryptoUtils decodeBase64:keyBase64];
    
    if (!dataBytes || !keyBytes) {
        return nil;
    }
    
    NSData *hmacBytes = [self computeHmac:dataBytes key:keyBytes algorithm:algorithm];
    if (hmacBytes) {
        return [CryptoUtils encodeBase64:hmacBytes];
    }
    return nil;
}

+ (HMACAlgorithm)algorithmByName:(NSString *)name {
    NSString *upperName = [name uppercaseString];
    
    if ([upperName isEqualToString:@"SHA1"] || [upperName isEqualToString:@"HMACSHA1"]) {
        return HMACAlgorithmSHA1;
    } else if ([upperName isEqualToString:@"SHA224"] || [upperName isEqualToString:@"HMACSHA224"]) {
        return HMACAlgorithmSHA224;
    } else if ([upperName isEqualToString:@"SHA256"] || [upperName isEqualToString:@"HMACSHA256"]) {
        return HMACAlgorithmSHA256;
    } else if ([upperName isEqualToString:@"SHA384"] || [upperName isEqualToString:@"HMACSHA384"]) {
        return HMACAlgorithmSHA384;
    } else if ([upperName isEqualToString:@"SHA512"] || [upperName isEqualToString:@"HMACSHA512"]) {
        return HMACAlgorithmSHA512;
    }
    
    return -1; // Unsupported algorithm
}

+ (NSArray<NSString *> *)supportedAlgorithms {
    return @[@"SHA1", @"SHA224", @"SHA256", @"SHA384", @"SHA512"];
}

@end

File: android/src/main/java/com/rocketchat/mobilecrypto/algorithms/HMACCrypto.kt
package com.rocketchat.mobilecrypto.algorithms

import java.security.InvalidKeyException
import java.security.NoSuchAlgorithmException
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * HMAC (Hash-based Message Authentication Code) operations
 */
object HMACCrypto {

    /**
     * Supported HMAC algorithms
     */
    enum class Algorithm(val algorithmName: String) {
        HMAC_SHA1("HmacSHA1"),
        HMAC_SHA224("HmacSHA224"),
        HMAC_SHA256("HmacSHA256"),
        HMAC_SHA384("HmacSHA384"),
        HMAC_SHA512("HmacSHA512")
    }

    /**
     * Compute HMAC using specified algorithm
     * 
     * @param data Data to authenticate
     * @param key Secret key
     * @param algorithm HMAC algorithm to use
     * @return HMAC bytes
     */
    @Throws(NoSuchAlgorithmException::class, InvalidKeyException::class)
    fun computeHmac(data: ByteArray, key: ByteArray, algorithm: Algorithm): ByteArray {
        val mac = Mac.getInstance(algorithm.algorithmName)
        val secretKey = SecretKeySpec(key, algorithm.algorithmName)
        mac.init(secretKey)
        return mac.doFinal(data)
    }

    /**
     * Compute HMAC-SHA256 with hex input/output
     * 
     * @param dataHex Data in hexadecimal format
     * @param keyHex Key in hexadecimal format
     * @return HMAC result in hexadecimal format
     */
    @Throws(NoSuchAlgorithmException::class, InvalidKeyException::class)
    fun hmacSha256Hex(dataHex: String, keyHex: String): String {
        val dataBytes = CryptoUtils.hexToBytes(dataHex)
        val keyBytes = CryptoUtils.hexToBytes(keyHex)
        val hmacBytes = computeHmac(dataBytes, keyBytes, Algorithm.HMAC_SHA256)
        return CryptoUtils.bytesToHex(hmacBytes)
    }

    /**
     * Compute HMAC with Base64 input/output
     * 
     * @param dataBase64 Base64-encoded data
     * @param keyBase64 Base64-encoded key
     * @param algorithm HMAC algorithm to use
     * @return Base64-encoded HMAC result
     */
    @Throws(NoSuchAlgorithmException::class, InvalidKeyException::class)
    fun hmacBase64(dataBase64: String, keyBase64: String, algorithm: Algorithm): String {
        val dataBytes = CryptoUtils.decodeBase64(dataBase64)
        val keyBytes = CryptoUtils.decodeBase64(keyBase64)
        val hmacBytes = computeHmac(dataBytes, keyBytes, algorithm)
        return CryptoUtils.encodeBase64NoWrap(hmacBytes)
    }

    /**
     * Compute HMAC with UTF-8 string input
     * 
     * @param data UTF-8 string data
     * @param keyBase64 Base64-encoded key
     * @param algorithm HMAC algorithm to use
     * @return Base64-encoded HMAC result
     */
    @Throws(NoSuchAlgorithmException::class, InvalidKeyException::class)
    fun hmacStringBase64(data: String, keyBase64: String, algorithm: Algorithm): String {
        val dataBytes = CryptoUtils.stringToUtf8Bytes(data)
        val keyBytes = CryptoUtils.decodeBase64(keyBase64)
        val hmacBytes = computeHmac(dataBytes, keyBytes, algorithm)
        return CryptoUtils.encodeBase64NoWrap(hmacBytes)
    }

    /**
     * Get algorithm by name string
     * 
     * @param name Algorithm name (case-insensitive)
     * @return Algorithm enum value
     * @throws IllegalArgumentException if algorithm not supported
     */
    fun getAlgorithmByName(name: String): Algorithm {
        return when (name.uppercase()) {
            "SHA1", "HMACSHA1" -> Algorithm.HMAC_SHA1
            "SHA224", "HMACSHA224" -> Algorithm.HMAC_SHA224
            "SHA256", "HMACSHA256" -> Algorithm.HMAC_SHA256
            "SHA384", "HMACSHA384" -> Algorithm.HMAC_SHA384
            "SHA512", "HMACSHA512" -> Algorithm.HMAC_SHA512
            else -> throw IllegalArgumentException("Unsupported HMAC algorithm: $name")
        }
    }
}
